components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,178 @@
+                writer.write(pingPath)
+                writer.newLine()
+                writer.write(ping)
+                writer.newLine()
+                writer.flush()
+                writer.close()
+            } catch (e: IOException) {
+                logger.warn("IOException while writing event to disk", e)
+            } finally {
+                stream?.close()
+            }
+        }
+    }
+
+    /**
+     * Function used by [PingUploadWorker] to process all serialized ping files for a particular
+     * ping type
+     *
+     * @param pingType String representing the ping type: baseline, events, metrics (or custom ping)
+     * @param callback Callback function to do the actual process action on the ping.  Typically
+     *                 this will be [HttpPingUploader.upload]
+     * @return Boolean representing the success of the upload task. This may be the value bubbled up
+     *         from the callback, or a local value if there was an error reading the files. The
+     *         [PingUploadWorker] uses this to determine whether or not to retry or declare success.
+     */
+    fun process(pingType: String, callback: (String, String) -> Boolean): Boolean {
+        listPingFiles(pingType).forEach { file ->
+            val reader: FileReader? = null
+
+            try {
+                val bufferedReader = BufferedReader(FileReader(file))
+                val path = bufferedReader.readLine()
+                val serializedPing = bufferedReader.readLine()
+
+                val processed = serializedPing == null || callback(path, serializedPing)
+
+                if (processed) {
+                    file?.delete()
+                } else {
+                    // The callback couldn't process this file.
+                    return false
+                }
+            } catch (e: FileNotFoundException) {
+                // This shouldn't happen after we queried the directory.
+                logger.error("Could not find ping file ${file?.name}")
+            } catch (e: IOException) {
+                // Something is not right.
+                logger.error("IO Exception when reading file ${file?.name}")
+                return false
+            } finally {
+                reader?.close()
+            }
+        }
+
+        return true
+    }
+
+    /**
+     * Helper function to list all ping files for a particular ping type
+     *
+     * @param pingType String representing the ping type: baseline, events, metrics (or custom ping)
+     * @return An Array<File?> containing the list of ping files
+     */
+    @VisibleForTesting
+    internal fun listPingFiles(pingType: String): Array<File?> {
+        val pingStorageDirectory = File(storageDirectory, pingType)
+
+        val uuidPattern = Pattern.compile(FILE_PATTERN)
+
+        val uuidFilenameFilter = FilenameRegexFilter(uuidPattern)
+        return pingStorageDirectory.listFiles(uuidFilenameFilter) ?: return arrayOfNulls(0)
```

Comment:
Instead o returning an array of nulls, can we return a nullable array and check at call site?

Reply:
Certainly!  This was another artifact from duplicating what previous Telemetry was doing.  Much more kotlin-esqe to return a nullable Array<>.